### PR TITLE
Session number binding

### PR DESCRIPTION
**Pull request by AndyGura**

### CODE REVIEW

File: src/auction/src/assets.mo [hunk 1]
```diff
@@ -24,7 +24,7 @@ module {
 
     public func historyLength() : Nat = Vec.size(history);
 
-    public func register(n : Nat) {
+    public func register(n : Nat, sessionsCounter : Nat) {
```
Comment (timohanke) [inline src/auction/src/assets.mo:27]: Why do we have a sessionsCounter argument here for the register() function?

**Review by timohanke — COMMENTED** [review]
Comments below

File: src/auction/src/orders.mo [hunk 2]
```diff
@@ -464,8 +504,18 @@ module {
...
       };
 
+      List.toIter(assetIdSet)
+      |> Iter.map<(T.AssetId, ()), (T.AssetInfo, T.AssetId)>(_, func(aid, _) = (Vec.get(assets.assets, aid), aid))
+      |> (
+        switch (checkSessionNumberMismatch(_)) {
+          case (?aid) return #err(#placement({ error = #SessionNumberMismatch(aid); index = 0 }));
+          case (_) {};
+        }
+      );
```
Comment (timohanke) [inline src/auction/src/orders.mo:517]: This whole block and the function checkSessionNumberMismatch can be replaced with a simple for loop:
```
for (aid in List.toIter(assetIdSet)) {
  ..
  // can return from the function here if mismatch
};
```

File: src/icrc1_auction_api_mock.mo [hunk 3]
```diff
@@ -330,26 +331,34 @@ actor class Icrc1AuctionAPI(adminPrincipal_ : ?Principal) = self {
     return null;
   };
 
-  public shared query ({ caller }) func queryTokenBids(ledger : Principal) : async [(Auction.OrderId, Order)] = async switch (getAssetId(ledger)) {
-    case (?aid) {
-      U.unwrapUninit(auction).getOrders(caller, #bid, ?aid)
-      |> Array.tabulate<(Auction.OrderId, Order)>(_.size(), func(i) = mapOrder(_ [i]));
+  public shared query ({ caller }) func queryTokenBids(ledger : Principal) : async ([(Auction.OrderId, Order)], Nat) {
+    let orders = switch (getAssetId(ledger)) {
```
Comment (timohanke) [inline src/icrc1_auction_api_mock.mo:335]: Isn't this switch better done with a let-else statement? Then we can return right away.

File: src/icrc1_auction_api_mock.mo [hunk 4]
```diff
@@ -330,26 +331,34 @@ actor class Icrc1AuctionAPI(adminPrincipal_ : ?Principal) = self {
...
+      case (?aid) {
+        U.unwrapUninit(auction).getOrders(caller, #bid, ?aid)
+        |> Array.tabulate<(Auction.OrderId, Order)>(_.size(), func(i) = mapOrder(_ [i]));
+      };
+      case (_) [];
     };
-    case (_) [];
+    (orders, U.unwrapUninit(auction).sessionsCounter);
```
Comment (timohanke) [inline src/icrc1_auction_api_mock.mo:342]: I don't really understand what `auction` is if `getAssetId(ledger)` returns null. How can that be defined?

**Review by timohanke — COMMENTED** [review]
Left some code comments.

Two more questions:
What is the `index` for that is inside the `#cancellation` error?

We have api, mock_api and api_new. Do we still need all that? Seems hard to maintain the code with repetition.

File: src/icrc1_auction_api_mock.mo (hunk 4 above)
Comment (AndyGura) [reply src/icrc1_auction_api_mock.mo:342]: The auction can be working, but user provided unknown principal as ledger principal

File: src/auction/src/orders.mo (hunk 2 above)
Comment (AndyGura) [reply src/auction/src/orders.mo]: done

File: src/icrc1_auction_api_mock.mo (hunk 3 above)
Comment (AndyGura) [reply src/icrc1_auction_api_mock.mo]: done

**Comment by timohanke** [comment]
What is the index field for that is inside the #cancellation error?

**Comment by AndyGura** [comment]
index field in #cancellation error response means:
- for cancellation `#orders : [{ #ask : OrderId; #bid : OrderId }]` means the index of provided order id, which can not be cancelled
- for cancellation `#all : ?[Principal]` index of principal which cannot be cancelled. Happens when provided principal is not registered as an ICRC1 ledger in the auction

File: src/auction/src/orders.mo [hunk 5]
```diff
@@ -346,12 +351,33 @@ module {
       switch (cancellations) {
         case (null) {};
         case (? #all(null)) {
+          switch (expectedSessionNumber) {
+            case (?sn) {
+              for ((asset, aid) in Vec.items(assets.assets)) {
+                if (asset.sessionsCounter != sn) {
+                  return #err(#cancellation({ error = #SessionNumberMismatch(aid); index = 0 }));
```
Comment (timohanke) [inline src/auction/src/orders.mo:358]: Why can we set the index to 0 here?

**Comment by timohanke** [comment]
Please do a search and replace for cancelation -> cancellation. In some places the word has a single `l`.

**Comment by AndyGura** [comment]
> Should we change the hierarchy of the errors and make #SessionMismatch a top-level error, not a sub-error of #cancellation and #placement errors?

Makes sense, done

> replace for cancelation -> cancellation
Done

File: src/auction/src/orders.mo (hunk 5 above)
Comment (AndyGura) [reply src/auction/src/orders.mo]: No more index in SessionNumberMismatch error

**Review by timohanke — APPROVED** [review]

**Pull request merged**